File: csharp/glp_link/primitives/LinkTerms.cs
using GlpRuntime.Link.Seam;
using GlpRuntime.Runtime;

namespace GlpRuntime.Link.Primitives;

/// <summary>
/// Maps between GLP ground terms and the host link value-types. The GLP surface is
/// <c>LinkId ::= link_id(Scheme, Endpoint, Nonce)</c>,
/// <c>Endpoint ::= String ; ep(String, Integer)</c>,
/// <c>Nonce ::= Integer ; String</c>, <c>Role ∈ {listener, connector}</c>, and the
/// fault lattice <c>ok | closed(LinkId,Reason) | tempFail(LinkId,Reason) |
/// permFail(LinkId,Reason)</c> (contracts/link-primitives.md §1, rulings-log
/// "fault term closed(LinkId,Reason); graceful reason eos").
/// </summary>
/// <remarks>
/// Parsing expects already-dereferenced GROUND terms — the GLP wrappers guard
/// <c>ground/1</c> before the kernel runs, so a malformed/unbound term reaching
/// here is a caller bug (it throws) rather than something to tolerate (CLAUDE.md
/// "robustness is a workaround in disguise").
/// </remarks>
public static class LinkTerms
{
    private const string LinkIdFunctor = "link_id";
    private const string EndpointFunctor = "ep";
    private const string RequestFunctor = "request";
    public const string GracefulReason = "eos";
    public const string AbruptReason = "abrupt";

    // ---- parse: term → host value ----

    /// <summary>
    /// Recursively dereference <paramref name="term"/> against the heap into a VarRef-free
    /// ground tree, so the <c>Parse*</c> helpers below see the bound constants at every depth.
    /// REQUIRED before parsing: the compiler represents a ground <c>link_id(Scheme, …)</c> with
    /// each struct ARG as a <see cref="VarRef"/> into a bound cell, so a shallow
    /// <c>heap.Dereference</c> resolves only the outer struct and leaves <c>Scheme</c>/<c>Endpoint</c>
    /// as VarRefs (the xUnit kernels passed ground <see cref="ConstTerm"/> args directly, hiding this).
    /// The GLP wrappers guard <c>ground/1</c>, so every nested cell is bound; an unbound cell at any
    /// depth is a caller bug → <see cref="ArgumentException"/> (the kernel aborts).
    /// </summary>
    public static Term GroundResolve(HeapFCP heap, Term term)
    {
        Term d = heap.Dereference(term);
        switch (d)
        {
            case ConstTerm c:
                return c;
            case StructTerm s:
                var args = new Term[s.Args.Count];
                for (int i = 0; i < s.Args.Count; i++)
                    args[i] = GroundResolve(heap, s.Args[i]);
                return new StructTerm(s.Functor, args);
            case VarRef v:
                throw new ArgumentException(
                    $"unbound cell {v.Addr} in a term expected ground (the ground/1 guard should have excluded it)");
            default:
                throw new ArgumentException($"cannot ground-resolve {d.GetType().Name}");
        }
    }

    /// <summary>Parse a ground <c>link_id(Scheme, Endpoint, Nonce)</c> term.</summary>
    public static LinkId ParseLinkId(Term term)
    {
        if (term is not StructTerm st || st.Functor != LinkIdFunctor || st.Args.Count != 3)
            throw new ArgumentException($"expected link_id/3 struct, got {Describe(term)}");
        var scheme = LinkScheme.Of(ConstString(st.Args[0], "LinkId.Scheme"));
        var endpoint = ParseEndpoint(st.Args[1]);
        var nonce = ParseNonce(st.Args[2]);
        return new LinkId(scheme, endpoint, nonce);
    }

    /// <summary>Parse a ground <c>Scheme</c> string (the first component of <c>rendezvous(Scheme, Endpoint)</c>).</summary>
    public static LinkScheme ParseScheme(Term term) => LinkScheme.Of(ConstString(term, "Scheme"));

    /// <summary>Parse a ground <c>Endpoint ::= String ; ep(String, Integer)</c> term (used by both LinkId and rendezvous).</summary>
    public static LinkAddress ParseEndpoint(Term term) => term switch
    {
        StructTerm ep when ep.Functor == EndpointFunctor && ep.Args.Count == 2 =>
            LinkAddress.Endpoint(ConstString(ep.Args[0], "ep host"), checked((int)ConstLong(ep.Args[1], "ep port"))),
        _ => LinkAddress.Path(ConstString(term, "Endpoint")),
    };

    private static LinkNonce ParseNonce(Term term)
    {
        var v = Const(term, "Nonce");
        return v switch
        {
            long l => LinkNonce.Int(l),
            int i => LinkNonce.Int(i),
            string s => LinkNonce.Str(s),
            _ => throw new ArgumentException($"Nonce must be Integer or String, got {v?.GetType().Name ?? "null"}"),
        };
    }

    /// <summary>
    /// Parse the ground close <c>Reason</c> (<c>Reason ::= String</c>): the atom <c>abrupt</c>
    /// from <c>link_close/1</c>, or a user reason from <c>link_close/2</c>. The wrapper
    /// grounds it, so a non-constant reaching here is a caller bug (it throws).
    /// </summary>
    public static string ParseReason(Term term) => ConstString(term, "Reason");

    /// <summary>Parse the ground establishment role atom (<c>listener</c>/<c>connector</c>).</summary>
    public static LinkRole ParseRole(Term term)
    {
        var s = ConstString(term, "Role");
        return s switch
        {
            "listener" => LinkRole.Listener,
            "connector" => LinkRole.Connector,
            _ => throw new ArgumentException($"Role must be listener|connector, got '{s}'"),
        };
    }

    // ---- build: host value → term ----

    /// <summary>
    /// Build the GLP <c>link_id(Scheme, Endpoint, Nonce)</c> term. String components
    /// (scheme, host, string nonce) are RE-QUOTED — the inverse of <see cref="Unquote"/> —
    /// so the reconstructed term is byte-identical to the GLP source literal
    /// <c>link_id("tcp", ep("127.0.0.1", 9100), 1)</c> the compiler produces (string
    /// constants carry their quotes). Without this, a path-B request token round-trip
    /// yields a bare <c>link_id(tcp, …)</c> that fails the <c>=?=</c> match in
    /// <c>accept_link</c> against the served (quoted) LinkId, and any program matching a
    /// <c>closed(link_id(…), …)</c> fault against a literal would likewise mismatch.
    /// </summary>
    public static Term ToTerm(LinkId id) => new StructTerm(LinkIdFunctor, new Term[]
    {
        new ConstTerm(Requote(id.Scheme.Name)),
        EndpointToTerm(id.Endpoint),
        NonceToTerm(id.Nonce),
    });

    private static Term EndpointToTerm(LinkAddress addr) => addr.Port is { } p
        ? new StructTerm(EndpointFunctor, new Term[] { new ConstTerm(Requote(addr.Host)), new ConstTerm((long)p) })
        : new ConstTerm(Requote(addr.Host));

    private static Term NonceToTerm(LinkNonce nonce) =>
        nonce.IsInteger ? new ConstTerm(nonce.IntValue) : new ConstTerm(Requote(nonce.StringValue!));

    // ---- in-band request token (T033 path-B handshake, OQ-A3) ----

    /// <summary>Build the in-band <c>request(LinkId, FromPeer)</c> handshake token.</summary>
    public static Term RequestToken(LinkId id, Term fromPeer) =>
        new StructTerm(RequestFunctor, new Term[] { ToTerm(id), fromPeer });

    /// <summary>Parse a ground <c>request(LinkId, FromPeer)</c> token (the listener's in-band first frame).</summary>
    public static (LinkId Id, Term FromPeer) ParseRequestToken(Term term)
    {
        if (term is not StructTerm st || st.Functor != RequestFunctor || st.Args.Count != 2)
            throw new ArgumentException($"expected request/2 token, got {Describe(term)}");
        return (ParseLinkId(st.Args[0]), st.Args[1]);
    }

    // ---- fault lattice terms (FR-043/045; rulings-log fault vocab) ----

    public static Term Ok() => new ConstTerm("ok");
    public static Term Closed(LinkId id, string reason) => Fault2("closed", id, reason);
    public static Term TempFail(LinkId id, string reason) => Fault2("tempFail", id, reason);
    public static Term PermFail(LinkId id, string reason) => Fault2("permFail", id, reason);

    /// <summary>Map a seam-level <see cref="LinkFaultSignal"/> to its GLP monitor term.</summary>
    public static Term FromSignal(LinkFaultSignal s) => s.Kind switch
    {
        LinkFaultKind.Closed => Closed(s.Link, s.Reason),
        LinkFaultKind.Transient => TempFail(s.Link, s.Reason),
        LinkFaultKind.Permanent => PermFail(s.Link, s.Reason),
        _ => throw new ArgumentOutOfRangeException(nameof(s)),
    };

    private static Term Fault2(string functor, LinkId id, string reason) =>
        new StructTerm(functor, new Term[] { ToTerm(id), new ConstTerm(reason) });

    // ---- const extraction helpers ----

    private static object? Const(Term term, string what) => term switch
    {
        ConstTerm c => c.Value is string s ? Unquote(s) : c.Value,
        VarRef => throw new ArgumentException($"{what}: unbound/unresolved VarRef — kernel must deep-deref before parsing"),
        _ => throw new ArgumentException($"{what}: expected an atomic constant, got {Describe(term)}"),
    };

    /// <summary>
    /// Strip the surrounding double-quotes a GLP STRING literal carries in its ConstTerm value.
    /// The compiler stores a string constant as <c>"value"</c> (quotes included) so the type
    /// checker can tell a String literal from a bare atom (parser.cs: "Wrap in quotes so type
    /// checker can distinguish strings from atoms"); a bare atom (e.g. <c>listener</c>,
    /// <c>abrupt</c>) carries no quotes. The kernels need the raw host value (e.g. the scheme
    /// token, the IPv4 host for <c>IPAddress.Parse</c>), so strip one surrounding pair if present.
    /// </summary>
    private static string Unquote(string s) =>
        s.Length >= 2 && s[0] == '"' && s[^1] == '"' ? s[1..^1] : s;

    /// <summary>Inverse of <see cref="Unquote"/>: wrap a raw host string value in the
    /// surrounding double-quotes a GLP STRING constant carries, so a term rebuilt from
    /// host data is structurally identical to the source literal (idempotent — a value
    /// that already carries quotes is left as-is).</summary>
    private static string Requote(string s) =>
        s.Length >= 2 && s[0] == '"' && s[^1] == '"' ? s : "\"" + s + "\"";

    private static string ConstString(Term term, string what) =>
        Const(term, what) as string ?? throw new ArgumentException($"{what}: expected a String constant, got {Describe(term)}");

    private static long ConstLong(Term term, string what) => Const(term, what) switch
    {
        long l => l,
        int i => i,
        _ => throw new ArgumentException($"{what}: expected an Integer constant, got {Describe(term)}"),
    };

    private static string Describe(Term term) => term switch
    {
        ConstTerm c => $"const({c.Value ?? "null"})",
        StructTerm s => $"{s.Functor}/{s.Args.Count}",
        VarRef v => $"var@{v.Addr}",
        _ => term.GetType().Name,
    };
}

File: glp_runtime/lib/link/primitives/link_terms.dart
import '../../runtime/heap_fcp.dart';
import '../../runtime/terms.dart';
import '../seam/link_address.dart';
import '../seam/link_fault.dart';
import '../seam/link_id.dart';
import '../seam/link_role.dart';
import '../seam/link_scheme.dart';

/// Maps between GLP ground terms and the host link value-types. The GLP surface is
/// `LinkId ::= link_id(Scheme, Endpoint, Nonce)`,
/// `Endpoint ::= String ; ep(String, Integer)`,
/// `Nonce ::= Integer ; String`, `Role ∈ {listener, connector}`, and the
/// fault lattice `ok | closed(LinkId,Reason) | tempFail(LinkId,Reason) |
/// permFail(LinkId,Reason)` (contracts/link-primitives.md §1, rulings-log
/// "fault term closed(LinkId,Reason); graceful reason eos").
///
/// Parsing expects already-dereferenced GROUND terms — the GLP wrappers guard
/// `ground/1` before the kernel runs, so a malformed/unbound term reaching
/// here is a caller bug (it throws) rather than something to tolerate (CLAUDE.md
/// "robustness is a workaround in disguise").
class LinkTerms {
  LinkTerms._();

  static const String _linkIdFunctor = 'link_id';
  static const String _endpointFunctor = 'ep';
  static const String _requestFunctor = 'request';
  static const String gracefulReason = 'eos';
  static const String abruptReason = 'abrupt';

  // ---- parse: term → host value ----

  /// Recursively dereference [term] against the heap into a VarRef-free
  /// ground tree, so the `Parse*` helpers below see the bound constants at every depth.
  /// REQUIRED before parsing: the compiler represents a ground `link_id(Scheme, …)` with
  /// each struct ARG as a [VarRef] into a bound cell, so a shallow
  /// `heap.dereference` resolves only the outer struct and leaves `Scheme`/`Endpoint`
  /// as VarRefs (the xUnit kernels passed ground [ConstTerm] args directly, hiding this).
  /// The GLP wrappers guard `ground/1`, so every nested cell is bound; an unbound cell at any
  /// depth is a caller bug → [ArgumentError] (the kernel aborts).
  static Term groundResolve(HeapFCP heap, Term term) {
    final Term d = heap.dereference(term);
    if (d is ConstTerm) {
      return d;
    } else if (d is StructTerm) {
      final args = List<Term>.filled(d.args.length, ConstTerm('nil'));
      for (int i = 0; i < d.args.length; i++) {
        args[i] = groundResolve(heap, d.args[i]);
      }
      return StructTerm(d.functor, args);
    } else if (d is VarRef) {
      throw ArgumentError(
          'unbound cell ${d.addr} in a term expected ground (the ground/1 guard should have excluded it)');
    } else {
      throw ArgumentError('cannot ground-resolve ${d.runtimeType}');
    }
  }

  /// Parse a ground `link_id(Scheme, Endpoint, Nonce)` term.
  static LinkId parseLinkId(Term term) {
    if (term is! StructTerm ||
        term.functor != _linkIdFunctor ||
        term.args.length != 3) {
      throw ArgumentError('expected link_id/3 struct, got ${_describe(term)}');
    }
    final scheme = LinkScheme.of(_constString(term.args[0], 'LinkId.Scheme'));
    final endpoint = parseEndpoint(term.args[1]);
    final nonce = _parseNonce(term.args[2]);
    return LinkId(scheme, endpoint, nonce);
  }

  /// Parse a ground `Scheme` string (the first component of `rendezvous(Scheme, Endpoint)`).
  static LinkScheme parseScheme(Term term) =>
      LinkScheme.of(_constString(term, 'Scheme'));

  /// Parse a ground `Endpoint ::= String ; ep(String, Integer)` term (used by both LinkId and rendezvous).
  static LinkAddress parseEndpoint(Term term) {
    if (term is StructTerm &&
        term.functor == _endpointFunctor &&
        term.args.length == 2) {
      return LinkAddress.endpoint(_constString(term.args[0], 'ep host'),
          _constLong(term.args[1], 'ep port'));
    }
    return LinkAddress.path(_constString(term, 'Endpoint'));
  }

  static LinkNonce _parseNonce(Term term) {
    final v = _const(term, 'Nonce');
    if (v is int) {
      return LinkNonce.int_(v);
    } else if (v is String) {
      return LinkNonce.str(v);
    } else {
      throw ArgumentError(
          'Nonce must be Integer or String, got ${v?.runtimeType ?? "null"}');
    }
  }

  /// Parse the ground close `Reason` (`Reason ::= String`): the atom `abrupt`
  /// from `link_close/1`, or a user reason from `link_close/2`. The wrapper
  /// grounds it, so a non-constant reaching here is a caller bug (it throws).
  static String parseReason(Term term) => _constString(term, 'Reason');

  /// Parse the ground establishment role atom (`listener`/`connector`).
  static LinkRole parseRole(Term term) {
    final s = _constString(term, 'Role');
    switch (s) {
      case 'listener':
        return LinkRole.listener;
      case 'connector':
        return LinkRole.connector;
      default:
        throw ArgumentError("Role must be listener|connector, got '$s'");
    }
  }

  // ---- build: host value → term ----

  /// Build the GLP `link_id(Scheme, Endpoint, Nonce)` term. String components
  /// (scheme, host, string nonce) are RE-QUOTED — the inverse of [_unquote] —
  /// so the reconstructed term is byte-identical to the GLP source literal
  /// `link_id("tcp", ep("127.0.0.1", 9100), 1)` the compiler produces (string
  /// constants carry their quotes). Without this, a path-B request token round-trip
  /// yields a bare `link_id(tcp, …)` that fails the `=?=` match in
  /// `accept_link` against the served (quoted) LinkId, and any program matching a
  /// `closed(link_id(…), …)` fault against a literal would likewise mismatch.
  static Term toTerm(LinkId id) => StructTerm(_linkIdFunctor, <Term>[
        ConstTerm(_requote(id.scheme.name)),
        _endpointToTerm(id.endpoint),
        _nonceToTerm(id.nonce),
      ]);

  static Term _endpointToTerm(LinkAddress addr) {
    final p = addr.port;
    return p != null
        ? StructTerm(_endpointFunctor,
            <Term>[ConstTerm(_requote(addr.host)), ConstTerm(p)])
        : ConstTerm(_requote(addr.host));
  }

  static Term _nonceToTerm(LinkNonce nonce) => nonce.isInteger
      ? ConstTerm(nonce.intValue)
      : ConstTerm(_requote(nonce.stringValue!));

  // ---- in-band request token (T033 path-B handshake, OQ-A3) ----

  /// Build the in-band `request(LinkId, FromPeer)` handshake token.
  static Term requestToken(LinkId id, Term fromPeer) =>
      StructTerm(_requestFunctor, <Term>[toTerm(id), fromPeer]);

  /// Parse a ground `request(LinkId, FromPeer)` token (the listener's in-band first frame).
  static (LinkId id, Term fromPeer) parseRequestToken(Term term) {
    if (term is! StructTerm ||
        term.functor != _requestFunctor ||
        term.args.length != 2) {
      throw ArgumentError('expected request/2 token, got ${_describe(term)}');
    }
    return (parseLinkId(term.args[0]), term.args[1]);
  }

  // ---- fault lattice terms (FR-043/045; rulings-log fault vocab) ----

  static Term ok() => ConstTerm('ok');
  static Term closed(LinkId id, String reason) => _fault2('closed', id, reason);
  static Term tempFail(LinkId id, String reason) =>
      _fault2('tempFail', id, reason);
  static Term permFail(LinkId id, String reason) =>
      _fault2('permFail', id, reason);

  /// Map a seam-level [LinkFaultSignal] to its GLP monitor term.
  static Term fromSignal(LinkFaultSignal s) {
    switch (s.kind) {
      case LinkFaultKind.closed:
        return closed(s.link, s.reason);
      case LinkFaultKind.transient:
        return tempFail(s.link, s.reason);
      case LinkFaultKind.permanent:
        return permFail(s.link, s.reason);
    }
  }

  static Term _fault2(String functor, LinkId id, String reason) =>
      StructTerm(functor, <Term>[toTerm(id), ConstTerm(reason)]);

  // ---- const extraction helpers ----

  static Object? _const(Term term, String what) {
    if (term is ConstTerm) {
      final v = term.value;
      return v is String ? _unquote(v) : v;
    } else if (term is VarRef) {
      throw ArgumentError(
          '$what: unbound/unresolved VarRef — kernel must deep-deref before parsing');
    } else {
      throw ArgumentError(
          '$what: expected an atomic constant, got ${_describe(term)}');
    }
  }

  /// Strip the surrounding double-quotes a GLP STRING literal carries in its ConstTerm value.
  /// The compiler stores a string constant as `"value"` (quotes included) so the type
  /// checker can tell a String literal from a bare atom (parser.cs: "Wrap in quotes so type
  /// checker can distinguish strings from atoms"); a bare atom (e.g. `listener`,
  /// `abrupt`) carries no quotes. The kernels need the raw host value (e.g. the scheme
  /// token, the IPv4 host for `IPAddress.Parse`), so strip one surrounding pair if present.
  static String _unquote(String s) =>
      s.length >= 2 && s[0] == '"' && s[s.length - 1] == '"'
          ? s.substring(1, s.length - 1)
          : s;

  /// Inverse of [_unquote]: wrap a raw host string value in the
  /// surrounding double-quotes a GLP STRING constant carries, so a term rebuilt from
  /// host data is structurally identical to the source literal (idempotent — a value
  /// that already carries quotes is left as-is).
  static String _requote(String s) =>
      s.length >= 2 && s[0] == '"' && s[s.length - 1] == '"' ? s : '"$s"';

  static String _constString(Term term, String what) {
    final v = _const(term, what);
    if (v is String) return v;
    throw ArgumentError(
        '$what: expected a String constant, got ${_describe(term)}');
  }

  static int _constLong(Term term, String what) {
    final v = _const(term, what);
    if (v is int) return v;
    throw ArgumentError(
        '$what: expected an Integer constant, got ${_describe(term)}');
  }

  static String _describe(Term term) {
    if (term is ConstTerm) {
      return 'const(${term.value ?? "null"})';
    } else if (term is StructTerm) {
      return '${term.functor}/${term.args.length}';
    } else if (term is VarRef) {
      return 'var@${term.addr}';
    } else {
      return term.runtimeType.toString();
    }
  }
}
